dev/src/cubegen.cpp

```diff
@@ -559,134 +559,160 @@ nfr("cg_load_vox", "name", "S", "R:voxels]S?",
+                        for (int i = 0; i < child_num; ++i) {
+                            int32_t child_node_id;
+                            if (ReadSpanInc(p, child_node_id))
+                                node_graph.insert_or_assign(child_node_id, node_id);
+                        }
                 } else if (!strncmp(id, "nSHP", 4)) {
-                    auto c = (const uint8_t*)p;
-                    auto node_id = ReadMemInc<int32_t>(c);
-                    auto dict_len = ReadMemInc<int32_t>(c);
-                    for (int i = 0; i<dict_len; ++i) {
+                    int32_t node_id;
+                    if (!ReadSpanInc(p, node_id)) return erreof();
+                    int32_t dict_len;
+                    if (!ReadSpanInc(p, dict_len)) return erreof();
+                    for (int i = 0; i < dict_len; ++i) {
                         string key;
-                        ReadVec<string, int32_t>(c, key);
-                        if (key == "_name") {
-                            string value;
-                            ReadVec<string, int32_t>(c, value);
-                            node_names.insert_or_assign(node_id, value);
-                        } else
-                            SkipVec<string, int32_t>(c);
+                        if (ReadSpanVec<string, int32_t>(p, key))
+                            if (key == "_name") {
+                                string value;
+                                if (ReadSpanVec<string, int32_t>(p, value))
+                                    node_names.insert_or_assign(node_id, value);
+                            } else
+                                SkipSpanVec<string, int32_t>(p);
                     }
-                    auto models_num = ReadMemInc<int32_t>(c);
-                    for (int i = 0; i<models_num; ++i) {
-                        auto model_id = ReadMemInc<int32_t>(c);
-                        node_to_model.insert_or_assign(node_id, model_id);
-                        auto dict_len = ReadMemInc<int32_t>(c);
-                        for (int i = 0; i<dict_len; ++i) {
-                            SkipVec<string, int32_t>(c);
-                            SkipVec<string, int32_t>(c);
+                    int32_t models_num;
+                    if (ReadSpanInc(p, models_num))
+                        for (int i = 0; i < models_num; ++i) {
+                            int32_t model_id;
+                            if (ReadSpanInc(p, model_id))
+                                node_to_model.insert_or_assign(node_id, model_id);
+                            int32_t dict_len;
+                            if (ReadSpanInc(p, dict_len))
+                                for (int i = 0; i < dict_len; ++i) {
+                                    SkipSpanVec<string, int32_t>(p);
+                                    SkipSpanVec<string, int32_t>(p);
+                                }
+                            [[maybe_unused]] int32_t reserved;
```

Comment:
neat use of feature, but it pushes complexity to user of the function.. better to default-init the 2nd arg in the function, and since that only happens on the error path, it should be cheap?

Reply:
alternatively, return a `pair<bool, T>`, and read with `auto [ok, t] =` though I suppose that doesn't allow convenient use inside `if`?

Reply:
yeah the result is used in if()return and && chainging in a few places in a way symmetric with the FileWrite methods